rfcs/theme-shape.md

```diff
@@ -0,0 +1,228 @@
+
+![Flamechart - Reduce number of tokens](theme-shape/perf-no-global.png)
+
+### Remove even more tokens
+
+It seems that the rendering performance depends on the number of CSS variables in the DOM. This is an experiment to remove even more variables to verify the hypothesis.
+
+- Removed alias shared color tokens, number of CSS variables reduced to 160.
+- **Each reflow takes ~20ms (350 ms total for the 10 mounts)**
+
+![Flamechart - Remove even more tokens](theme-shape/perf-no-colors.png)
+
+### Hash CSS variables
+
+In inline aliases scenario, there are 1200 CSS variables with names like `--alias-color-neutral-neutralStrokeAccessibleHover`. The total string length of CSS variable names (including the leading `--`) is 39,252 characters - is that affecting the perf as well? We used 7-char hashes for variable names to check for any difference - this reduced the string length to 10,354 characters. There is no difference in the performance.
+
+- **Each reflow takes ~90ms (1000 ms total for the 10 mounts)**
+
+**The perf example intentionally uses precomputed hashes to keep the hash computation out of the measured code path.**
+
+![Flamechart - Hash CSS variables](theme-shape/perf-hash-variables.png)
+
+### Results Summary
+
+We also used flamegrill to compare the approaches. Inlined aliases is used as a baseline all other approaches are compared to. For each experiment we run the test 10 times for both the experiment and the baseline to get comparable numbers.
+
+![Flamegrill - Results Summary](theme-shape/perf-flamegrill.png)
+
+## Proposed changes to the theme shape
+
+1. [Inline values in alias tokens](#inline-alias)
+2. [Do not inject global color tokens as CSS variables](#no-global-tokens)
+3. [Merge global and alias tokens to a single object](#merge-global-alias)
+4. [Flatten the tokens object](#flatten-tokens)
+5. [Hash variable names](#hash-variables)
+6. [Be cautious when adding new tokens to the theme](#be-cautious)
+7. [Get rid of all CSS variables](#no-css-vars)
+
+### <a name="inline-alias"></a>Inline values in alias tokens
+
+Global tokens reference values, alias tokens reference the global tokens:
+
+```css
+/* Before */
+--global-red: #f00;
+--alias-error: var(--global-red);
+```
+
+Performance analysis shows, that by inlining the values directly into CSS variables for alias tokens we can improve Rendering perf.
+
+Proposed change:
+
+```css
+/* After */
+--global-red: #f00; /* no change here */
+--alias-error: #f00; /* inlined value from --global-red */
+```
+
+This change has already been implemented in [#19660](https://github.com/microsoft/fluentui/pull/19660).
+
+### <a name="no-global-tokens"></a>Do not inject global color tokens as CSS variables
+
+The main reason why CSS variables are used is to be able to switch themes just by changing the CSS variables values.
+
+But, by design, global tokens never change their values when switching theme, so there is no point to have them as CSS variables.
```

Comment:
Isn't the brand palette considered a global? For my team (office.com) this IS what changes when the user changes themes, in addition to the dark/light/hc token assignments.

Reply:
I should mention we have a big release slated for Ignite and we are relying heavily on this theme switching mechanic. Preventing changing brand colors at runtime will be disruptive. If this has to happen, we'll need time to discuss a new theming strategy and I don't think we'll have time before Nov.

Reply:
It is true that today brand palette is considered global and changes on theme switching. I think this is confusing and complicates the mental model of tokens. It seems like brand colors should move out of global tokens, as far as I know from talking to daisy and @levithomason their existence as global tokens is more historical than for a specific purpose. 

In my opinion, it would be a far simpler model (both mental and actual) if all that changed on theme switch was alias tokens and global tokens were both global and effectively static at runtime. Perhaps global tokens should even be a design-time concept and stamped out as part of the token pipeline.

Clearly that is a significant change, we'd need to manage it so that we don't break folks (or help folks move over if it's easy) especially if we have Ignite timelines. 

Is there a particular reason for why it's desirable to have brand colors part of global tokens? Want to make sure I'm not missing something.

Reply:
> It is true that today brand palette is considered global and changes on theme switching.

I would not completely agree. Brand palette (`theme.global.palette.brand`) is expected to be global and never change when switching themes.

The intended usecase for `create*Theme` functions was you call them with your custom brand ramp but **use the same ramp for all light, dark and HC**.
![image](https://user-images.githubusercontent.com/9615899/133873405-9f85574c-a97d-43a5-bbcf-303d1f897f9d.png)

Then it is a responsibility of alias tokens to do appropriate mapping in different themes:
![image](https://user-images.githubusercontent.com/9615899/133873441-f380f8ea-0d5d-4d91-b5cd-1499f797e0f8.png)

Reply:
I will check the code you linked, but what is your motivation for using a different brand ramp in light vs dark?

Is it that the mapping in aliases does not match you design or do you use global color tokens directly?
We do not expect anybody to use the global color tokens and plan to remove them completely from the theme.

Let's have a meeting and discuss the impact this would have on your scenario.

Reply:
We don’t change the ramp between light and dark, but our app allows users and tenants to customize brand colors. You can try the theme switcher yourself if you visit www.office.com, open the settings tab in the header, and modify the selected theme. I’d love to have a chat if this usage is unexpected.